Dockerfile

```diff
@@ -1,20 +1,39 @@
-FROM ruby:2.7.5
-RUN apt-get update -qq && apt-get upgrade -y
+ARG base_image=ruby:2.7.5-slim-buster
+FROM $base_image AS builder
 
-RUN apt-get install -y build-essential nodejs && apt-get clean
+ENV RAILS_ENV=production
+# TODO: have a separate build image which already contains the build-only deps.
+RUN apt-get update -qy && \
+    apt-get upgrade -y && \
+    apt-get install -y build-essential nodejs && \ 
+    apt-get clean
 
-ENV GOVUK_APP_NAME feedback
-ENV PORT 3028
-ENV RAILS_ENV development
+RUN mkdir /app
+WORKDIR /app
+COPY Gemfile* .ruby-version /app/
+RUN bundle config set deployment 'true' && \
+    bundle config set without 'development test' && \
+    bundle install -j8 --retry=2
 
-ENV APP_HOME /app
-RUN mkdir $APP_HOME
+COPY . /app
+# TODO: We probably don't want assets in the image; remove this once we have a proper deployment process which uploads to (e.g.) S3.
+RUN GOVUK_WEBSITE_ROOT=https://www.gov.uk \
+    GOVUK_APP_DOMAIN=www.gov.uk \
+    bundle exec rails assets:precompile
 
-WORKDIR $APP_HOME
-ADD Gemfile* $APP_HOME/
-ADD .ruby-version $APP_HOME/
-RUN bundle install
 
-ADD . $APP_HOME
+FROM $base_image
 
-CMD bash -c "bundle exec rails s -p $PORT -b '0.0.0.0'"
+ENV RAILS_ENV=production GOVUK_APP_NAME=manuals-feedback
+
+RUN apt-get update -qy && \
```

Comment:
Could you explain why we go back to the base image and run apt-get install nodejs again? (Sorry, not expert enough on dockerfiles to get what the intent is here).

Reply:
I believe Nodejs is required to both install and run the Ruby app. In the base image, we include it to be able to install/compile assets. In the "child" image, we need it to run the Ruby app. 

There will be another pass to clean this Dockerfile later, we just need it to be an ok shape.

Reply:
Ah, so we make a first image as builder, install all the build requirements, do the build, then we clean the slate and make another image (the runtime image?) from the original base, install only what we need to run (not build), but then copy across the built app files from the builder image? So we end up with a smaller runtime image? 

Neat (if I'm understanding it correctly)

Reply:
yes, you are correct. In Ruby, you don't get as much benefit as with a compiled language like golang and c.

You can read some further doc here: https://docs.docker.com/develop/develop-images/multistage-build/

Reply:
Thanks!